File: backend/app/services/admin_service.py
"""Admin moderation, verification, and user management."""
from __future__ import annotations

from app.core.exceptions import NotFoundError
from app.models.enums import ApprovalStatus, JobStatus, UserStatus
from app.repositories.company_repo import CompanyRepository
from app.repositories.job_repo import JobRepository
from app.repositories.user_repo import (
    RecruiterProfileRepository,
    UserRepository,
)
from app.services.notification_service import NotificationService


class AdminService:
    def __init__(
        self,
        users: UserRepository,
        recruiters: RecruiterProfileRepository,
        companies: CompanyRepository,
        jobs: JobRepository,
        notifications: NotificationService,
    ) -> None:
        self.users = users
        self.recruiters = recruiters
        self.companies = companies
        self.jobs = jobs
        self.notifications = notifications

    async def list_users(self, *, role: str | None = None, limit: int = 50, skip: int = 0):
        query = {"role": role} if role else {}
        items = await self.users.find(query, sort=[("created_at", -1)], limit=limit, skip=skip)
        total = await self.users.count(query)
        return {"items": [u.model_dump(mode="json", exclude={"password_hash"}) for u in items],
                "total": total, "limit": limit, "skip": skip,
                "has_more": skip + len(items) < total}

    async def set_user_status(self, user_id: str, status: UserStatus):
        user = await self.users.set_status(user_id, status)
        if not user:
            raise NotFoundError("User not found.")
        return user

    async def pending_recruiters(self):
        return await self.recruiters.list_pending()

    async def verify_recruiter(self, profile_id: str, verified: bool = True):
        profile = await self.recruiters.update(profile_id, {"verified": verified})
        if not profile:
            raise NotFoundError("Recruiter profile not found.")
        await self.notifications.notify(
            profile.user_id,
            type_="recruiter.verified",
            title="Recruiter verified" if verified else "Verification declined",
            body="You can now post jobs." if verified else "Please resubmit your documents.",
        )
        return profile

    async def pending_companies(self):
        return await self.companies.list_pending()

    async def moderate_company(self, company_id: str, approve: bool):
        status = ApprovalStatus.APPROVED if approve else ApprovalStatus.REJECTED
        company = await self.companies.update(company_id, {"approval_status": status.value})
        if not company:
            raise NotFoundError("Company not found.")
        await self.notifications.notify(
            company.owner_id,
            type_="company.moderated",
            title=f"Company {status.value}",
            body=f"{company.name} was {status.value}.",
        )
        return company

    async def moderation_queue(self):
        return await self.jobs.find(
            {"status": JobStatus.PENDING_MODERATION.value}, sort=[("created_at", 1)], limit=100
        )

    async def moderate_job(self, job_id: str, approve: bool):
        status = JobStatus.PUBLISHED if approve else JobStatus.REJECTED
        job = await self.jobs.update(job_id, {"status": status.value})
        if not job:
            raise NotFoundError("Job not found.")
        await self.notifications.notify(
            job.posted_by,
            type_="job.moderated",
            title=f"Job {status.value}",
            body=f"'{job.title}' was {status.value}.",
            link=f"/recruiter/jobs/{job.id}",
        )
        return job

File: backend/src/services/adminService.js
"use strict";

/** Admin moderation, verification, and user management.
 * Mirrors app/services/admin_service.py. */

const { User, RecruiterProfile, Company, Job } = require("../models");
const { ApprovalStatus, JobStatus } = require("../models/enums");
const notificationService = require("./notificationService");
const { NotFoundError } = require("../utils/errors");
const { adminUser, page } = require("../utils/serialize");

async function listUsers({ role = null, limit = 50, skip = 0 } = {}) {
  const query = role ? { role } : {};
  const items = await User.find(query).sort({ created_at: -1 }).skip(skip).limit(limit);
  const total = await User.countDocuments(query);
  return page(items.map(adminUser), total, limit, skip);
}

async function setUserStatus(userId, status) {
  const user = await User.findByIdAndUpdate(userId, { $set: { status } }, { new: true });
  if (!user) throw new NotFoundError("User not found.");
  return user;
}

async function pendingRecruiters() {
  return RecruiterProfile.find({ verified: false }).limit(50);
}

async function verifyRecruiter(profileId, verified = true) {
  const profile = await RecruiterProfile.findByIdAndUpdate(
    profileId,
    { $set: { verified } },
    { new: true }
  );
  if (!profile) throw new NotFoundError("Recruiter profile not found.");
  await notificationService.notify(profile.user_id, {
    type: "recruiter.verified",
    title: verified ? "Recruiter verified" : "Verification declined",
    body: verified ? "You can now post jobs." : "Please resubmit your documents.",
  });
  return profile;
}

async function pendingCompanies() {
  return Company.find({ approval_status: ApprovalStatus.PENDING })
    .sort({ created_at: 1 })
    .limit(50);
}

async function moderateCompany(companyId, approve) {
  const status = approve ? ApprovalStatus.APPROVED : ApprovalStatus.REJECTED;
  const company = await Company.findByIdAndUpdate(
    companyId,
    { $set: { approval_status: status } },
    { new: true }
  );
  if (!company) throw new NotFoundError("Company not found.");
  await notificationService.notify(company.owner_id, {
    type: "company.moderated",
    title: `Company ${status}`,
    body: `${company.name} was ${status}.`,
  });
  return company;
}

async function moderationQueue() {
  return Job.find({ status: JobStatus.PENDING_MODERATION }).sort({ created_at: 1 }).limit(100);
}

async function moderateJob(jobId, approve) {
  const status = approve ? JobStatus.PUBLISHED : JobStatus.REJECTED;
  const job = await Job.findByIdAndUpdate(jobId, { $set: { status } }, { new: true });
  if (!job) throw new NotFoundError("Job not found.");
  await notificationService.notify(job.posted_by, {
    type: "job.moderated",
    title: `Job ${status}`,
    body: `'${job.title}' was ${status}.`,
    link: `/recruiter/jobs/${job.id}`,
  });
  return job;
}

module.exports = {
  listUsers,
  setUserStatus,
  pendingRecruiters,
  verifyRecruiter,
  pendingCompanies,
  moderateCompany,
  moderationQueue,
  moderateJob,
};
